src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs

```diff
@@ -3,106 +3,126 @@
 // See the LICENSE file in the project root for more information
 
 using System;
-using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.Linq;
+using System.Text.Encodings.Web;
+using System.Text.Json;
 using System.Threading;
-using Elastic.CommonSchema;
+using System.Threading.Channels;
+using System.Threading.Tasks;
 using Elasticsearch.Net;
 
 namespace Elasticsearch.Extensions.Logging
 {
 	internal class ElasticsearchDataShipper : IDisposable
 	{
-		private const int MaxQueuedMessages = 1024;
+		private readonly List<Task<Task>> _backgroundTasks =  new List<Task<Task>>();
 
-		private readonly BlockingCollection<LogEvent> _messageQueue =
-			new BlockingCollection<LogEvent>(MaxQueuedMessages);
-
-		private readonly Thread _outputThread;
-
-		public ElasticsearchDataShipper()
+		public ElasticsearchDataShipper(ElasticsearchLoggerOptions options)
 		{
-			_outputThread = new Thread(ProcessLogQueue) { IsBackground = true, Name = $"{nameof(ElasticsearchLogger)}.{nameof(ProcessLogQueue)}" };
-			_outputThread.Start();
+			_options = options;
+			var maxConsumers = Math.Max(1, options.Throttles.ConcurrentConsumers);
+			PubSub = Channel.CreateBounded<LogEvent>(new BoundedChannelOptions(options.Throttles.MaxInFlightMessages)
+			{
+				SingleReader = maxConsumers == 1,
+				AllowSynchronousContinuations = true,
+				// wait does not block it simply signals that Writer.TryWrite should return false and be retried
+				// DropWrite will make `TryWrite` always return true, which is not what we want.
+				FullMode = BoundedChannelFullMode.Wait
+			});
+			async Task ConsumeMessages() =>
```

Comment:
I think `ConsumeMessages` could be removed and `Consume` called directly in `_backgroundTasks.Add` taking `Func<Task>` on line 36.

Reply:
++ left over from debugging

Reply:
I was thinking along the lines of

```c#
private readonly List<Task> _backgroundTasks =  new List<Task>();
```

```c#
for (var i = 0; i < maxConsumers; i++)
    _backgroundTasks.Add(Task.Factory.StartNew(() => Consume(options), TaskCreationOptions.LongRunning));
```

with `Consume` taking `ElasticsearchLoggerOptions` so that changes to options will affect consumers _(might be unexpected if changes to `options.Throttles` don't affect existing consumers)_.